### PR TITLE
DCB-130 Extend Application metadata with preferred field

### PR DESCRIPTION
**Pull request by lukasz-andrzejak**

### CODE REVIEW

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationPreferredHelper.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+public class ApplicationPreferredHelper {
+
+    private ApplicationPreferredHelper() {
+    }
+
+    public static List<Record10<String, String, String, String, String, String, Integer, String, JSONB, Boolean>>
+    collectPreferredVersionFirstListStb(Result<Record10<String, String, String, String, String, String, Integer, String, JSONB, Boolean>> result) {
+        return result.stream()
+                .filter(record1 -> result.stream()
+                        .noneMatch(record2 -> record1 != record2
```
Comment (den314) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationPreferredHelper.java:43]: record1 != record2 - can be moved to .filter

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationPreferredHelper.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    public static List<Record13<String, String, String, String, String, Boolean, Boolean, Boolean, String, String, Integer, String, JSONB>>
+    collectPreferredVersionFirstForListMaintainer(Result<Record13<String, String, String, String, String, Boolean, Boolean, Boolean, String, String, Integer, String, JSONB>> result) {
+        return result.stream()
+                .filter(record1 -> result.stream()
+                        .noneMatch(record2 -> record1 != record2
+                                && record1.get(APPLICATION.ID_RDOMAIN).equals(record2.get(APPLICATION.ID_RDOMAIN))
+                                && record2.get(APPLICATION.PREFERRED).equals(true))
+                ).collect(Collectors.toList());
```
Comment (den314) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationPreferredHelper.java:56]: seems collectPreferredVersionFirstListStb and collectPreferredVersionFirstForListMaintainer share the same logic, let's put it to single method and re-use

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationPreferredHelper.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+import static com.lgi.appstore.metadata.jooq.model.tables.Application.APPLICATION;
+
+public class ApplicationPreferredHelper {
+
+    private ApplicationPreferredHelper() {
+    }
+
+    public static List<Record10<String, String, String, String, String, String, Integer, String, JSONB, Boolean>>
+    collectPreferredVersionFirstListStb(Result<Record10<String, String, String, String, String, String, Integer, String, JSONB, Boolean>> result) {
```
Comment (den314) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationPreferredHelper.java:40]: collectPreferredVersionFirstListStb -> matchByPreferredVersionAndDomain (or similar), also for collectPreferredVersionFirstForListMaintainer

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationPreferredHelper.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                                && record1.get(APPLICATION.ID_RDOMAIN).equals(record2.get(APPLICATION.ID_RDOMAIN))
+                                && record2.get(APPLICATION.PREFERRED).equals(true))
+                ).collect(Collectors.toList());
+    }
+
+    public static Optional<Record20<String, String, String, String, String, String, String, String, String, String, String, String, JSONB, JSONB, JSONB, JSONB, JSONB, Integer, String, Boolean>>
+    collectPreferredVersionFirstForDetailsStb(Result<Record20<String, String, String, String, String, String, String, String, String, String, String, String, JSONB, JSONB, JSONB, JSONB, JSONB, Integer, String, Boolean>> result) {
+        if (result.size() == 1) {
+            return Optional.of(result.get(0));
+        } else if (result.size() > 1) {
```
Comment (den314) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationPreferredHelper.java:63]: seems if size == 1 isn't needed, if stream contains 1 value findAny() will still return the only value from the stream. Same in other method

File: appstore-metadata-service-tests/src/test/groovy/com/lgi/appstore/metadata/test/cases/real/sanity/DevApiFTSpecSanity.groovy [hunk 5]
```diff
@@ -173,6 +177,7 @@ class DevApiFTSpecSanity extends AsmsSanitySpecBase {
         field().header().version().from(theBody1) == v1
         field().header().visible().from(theBody1) == v1Visible
         field().header().encryption().from(theBody1) == v1Encryption
+        field().header().preferred().from(theBody1) == v1Preferred
```
Comment (den314) [inline appstore-metadata-service-tests/src/test/groovy/com/lgi/appstore/metadata/test/cases/real/sanity/DevApiFTSpecSanity.groovy:180]: if there's no test for retrieval app by preferred field, would be great to add one

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationPreferredHelper.java (hunk 2 above)
Comment (lukasz-andrzejak) [reply appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationPreferredHelper.java:56]: @den314 I tried it, but there is a problem with strong typing Records type. The method needs exactly RecordX where X is a number.

File: appstore-metadata-service-tests/src/test/groovy/com/lgi/appstore/metadata/test/cases/real/sanity/DevApiFTSpecSanity.groovy (hunk 5 above)
Comment (lukasz-andrzejak) [reply appstore-metadata-service-tests/src/test/groovy/com/lgi/appstore/metadata/test/cases/real/sanity/DevApiFTSpecSanity.groovy:180]: Added

**Review by den314 — APPROVED** [review]

**Pull request merged**